### ISSUE TITLE
Unspecced 'sender' field on presence events in /sync

### ISSUE THREAD

**Issue by joepie91** [matrix-org/matrix-doc#2680]
The [/sync spec](https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-client-r0-sync) claims:

- The `presence` property is of type `Presence`
- The `Presence` type has an `events` property that contains an array of `Event`-type items
- The `Event` type has `content` and `type` fields

However, in practice, the presence events also contain a `sender` field, which is not accounted for in the spec:

```js
// ...
	"presence": {
		"events": [{
			"type": "m.presence",
			"sender": "@redacted1:example.com",
			"content": {
				"presence": "offline",
				"last_active_ago": 55595
			}
		}, {
			"type": "m.presence",
			"sender": "@redacted2:example.com",
			"content": {
				"presence": "offline",
				"last_active_ago": 38143
			}
		}, 
// ...
```

This `sender` property *is* [specified for the `to_device` events](https://matrix.org/docs/spec/client_server/r0.6.1#extensions-to-sync), but that has no relation to `presence`.

**Comment by uhoreg** [comment]
Similarly, the `Ephemeral` type has an `events` property of type `[Event]`, but in the example given, the ephemeral event has a `room_id` field.  (I haven't checked if the example matches what synapse sends.)

It looks like we either need to split `Event` into `Ephemeral Event`, `Account Data Event`, and `Presence Event`, or make `room_id` and `sender` optional fields in `Event`.  Splitting into three types is probably the clearer way to do it, although it make the spec longer.

**Comment by joepie91** [comment]
@uhoreg Looking at #2681 as well, I feel like maybe the spec would benefit from generally restructuring its documentation around event formats. Specify each format as a distinct format (Room Event, State Event, Ephemeral Event, Presence Event, Account Data Event, Stripped Event, Device Event, ...) and then clearly identify which are subtypes of which others, and therefore may appear anywhere the more generic type is expected.

The list of and relation between these types could then probably be captured in a single tree, like so:

```
Event
|- Ephemeral Event
|- Device Event
|- Account Data Event
|- Presence Event
|- Room Event
|  `- State Event
`- Stripped Event
```

(Note that I'm not sure whether that tree is actually *correct*, it's just to illustrate the idea.)

I think that even when dedicating a section to *all* different event types and splitting out `Event` into 3 distinct types, we could still end up with a shorter spec in the end - by removing the current inline definitions of `State Event` and `Room Event` in the `/sync` endpoint, and making those hyperlinks to the "Event types" section instead (which would be a more generic version of the current "9.1 Types of room events" section).

That would reduce duplication and also make sure that someone looking at the `/sync` documentation will be presented with the more general information about how different event types relate to each other, which is currently a rough spot.

Thoughts?

**Comment by joepie91** [comment]
To motivate my reasoning for the above suggestion of "centralized type definitions" a little further, since I just realized that it might not be obvious to everyone:

Currently, the situation is that type definitions are somewhat spread throughout the spec. Assuming there are roughly two demographics of implementors, those who implement the spec partially for a specific usecase and those who implement it fully, that means:
- __Partial implementors:__ can implement a specific bit of the spec directly from its section of the spec, without navigating further.
- __Full implementors:__ need to dig around in the specification a lot to cobble together all the different types, and form a coherent model of the full API's types.

With centralized type definitions, this would change to:
- __Partial implementors:__ can implement a specific bit of the spec from its section of the spec *and by following the type definition hyperlinks*.
- __Full implementors:__ can linearly read the type definitions and build a mental model out of them, without having to dig through the spec for that.

In other words, it would introduce slightly more navigation for partial implementors, but *significantly* reduce work (and implementation error potential) for full implementors.

**Comment by KitsuneRal** [comment]
As an author of a client application and library that aspire to implement the whole spec - a Full Implementor, as you called it - I still very much appreciate that I don't need to jump across the spec in the search of centralised definitions (I have to do it at cppreference.com, so I can compare - and not that I like it). However, I used to (and continue to) spend quite a bit of time looking into the _source_, that is [the formal API definitions](https://github.com/matrix-org/matrix-doc/tree/master/api) and [event schemas](https://github.com/matrix-org/matrix-doc/tree/master/event-schemas). These _are_ mostly normalised and only have a very small bit of duplication. Reading OpenAPI definitions directly is not hard, and definitely gives a very good sense of how Matrix data structures relate to each other. My personal take on this was to further [automatically generate](https://github.com/KitsuneRal/gtad) C++ code from the OpenAPI definitions so that I could have all of them ready for use right in my IDE; but I recognise that very few would go that far.

**Comment by KitsuneRal** [comment]
With that said, the event type hierarchy you described above would certainly be a worthwhile addition to the spec; the current hierarchy is lacking, and the fact that we don't formally define, e.g., ephemeral events as first-class citizens next to, e.g., state events has been a matter of confusion before.
